compiler/qsc_frontend/src/typeck/rules.rs

```diff
@@ -328,11 +328,22 @@ impl<'a> Context<'a> {
             }
             ExprKind::Lambda(kind, input, body) => {
                 let input = self.infer_pat(input);
-                let body = self.infer_expr(body).ty;
+                let prev_ret_ty = self.return_ty.take();
+                let output_ty = self.inferrer.fresh_ty(TySource::not_divergent(body.span));
+                self.return_ty = Some(output_ty);
+                let body_ty = self.infer_expr(body).ty;
+                let output_ty = self
+                    .return_ty
+                    .take()
+                    .expect("return type should be present");
+                self.return_ty = prev_ret_ty;
+                if body_ty != Ty::UNIT {
+                    self.inferrer.eq(body.span, output_ty.clone(), body_ty);
```

Comment:
Why is if block needed? What does it do exactly? (Maybe add a comment for the noobs like me).

Reply:
Added clarifying comment